client/src/controllers/SwapController.ts

```diff
@@ -173,23 +189,50 @@ export class SwapController extends Controller<
   }
 
   /**
-   * Perform a search based on a single input query, and only if that query's value
-   * differs from the current matching URL param. Once complete, update the URL param.
-   * Additionally, clear the `'p'` pagination param in the URL if present, can be overridden
-   * via action params if needed.
+   * Perform a search either based on a single input query or the serialised input's form values.
+   *
+   * By default, it will perform a request based on the input's form values serialised.
+   * Do not account for anything in the main location/URL, simply replace the content within
+   * the target element.
+   *
+   * If the `location` param is set to true, then the search will only be performed based on the
+   * input and comparison against the matching URL query param. Additionally, cleared params will
+   * be removed from the URL if present. `clear` and `location` can be provided as Event detail
+   * or action params.
    */
   search(
-    data?: CustomEvent<{ clear: string }> & {
-      params?: { clear?: string };
+    data?: CustomEvent<{ clear: string; location: boolean }> & {
+      params?: { clear?: string; location: boolean };
     },
   ) {
+    if (!(data?.detail || data?.params || {}).location) {
+      const form = (
+        this.hasInputTarget ? this.inputTarget.form : this.element
+      ) as HTMLFormElement;
+
+      // serialise the form to a query string
+      // https://github.com/microsoft/TypeScript/issues/43797
+      const searchParams = new URLSearchParams(new FormData(form) as any);
+
+      const queryString = '?' + searchParams.toString();
+      const url = this.srcValue;
+
+      this.replace(url + queryString);
+
+      return;
```

Comment:
100% agree, I think that the action Params approach does feel wrongh here and discrete methods make sense. Just some suggestions on naming so that the existing header search implementation does not need to change. 

How about...
- `search`/`searchLazy` for the location based query/update (no change to current)
- `submit`/`submitLazy` for the form serialisation based behaviour that this PR will add
- `request` / add `requestLazy` - using requestLazy is not needed now but gives a nice consistency to the approach

This way it's more explicit, simpler names and sets us up for a way of thinking about this Controller as Async swapping in of content.

Also, this means we don't have to think about 'default' behaviour being changed with values or Params but rather you just call the action/method combo that makes sense for what you need. Works without needing to change anything on how the now merged header search usage works.

Let me know if this approach is sound, I'll update next week when I'm back from holidays. Hopefully I've understood your comments correctly. 

As for your follow-up question - yes, action params are shared across all events attached via actions on the same DOM element. They're super useful for cases where you want different data but a value does not make sense.

Reply:
@laymonage  I have now updated the PR with the approach explained above and added unit tests for calling `replaceLazy` directly also.